ui: unsafe lifecycle methods

vladlos opened the pull request:
changed `componentWillMount` to `componentDidMount` and `componentWillReceiveProps` to `componentDidUpdate`.
fixed broken code parts due to change lifecycle methods

Resolves: #45189

Release justification: bug fixes and low-risk updates to new functionality

Release note (ui): none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45947)
<!-- Reviewable:end -->

dhartunian reviewed (changes requested):
> ui: unsafe lifecycle methods

this title doesn't clearly communicate what is being changed

> changed componentWillMount to componentDidMount and componentWillReceiveProps to componentDidUpdate.

can you clarify if this changes any behavior in the components? link to relevant react documentation.

> fixed broken code parts due to change lifecycle methods

every change needs reasoning here. it's not obvious from the code changes.

> Reviewed 33 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-:-M2A4CVy9rBF8AF0fyUr:b-ao6rd2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster)

___
*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-M2A4QA5APn_7XqWtpX2:-M2A4QA5APn_7XqWtpX3:b-t2y4kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/11449b96430bde899bc58c028df6cde9f0fc2807/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L168)):*
> ```XML
>   componentDidUpdate() {
>     const zoomTransform = this.state.zoomTransform.withViewportSize(this.props.viewportSize);
>     if (!_.isEqual(this.state.zoomTransform, zoomTransform)) {
> ```

Why is this necessary to add? It's unclear what this code is doing. Can a function be extracted that names the reason?
___
*[pkg/ui/src/views/app/containers/timewindow/index.tsx, line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-M2A4y2u-gKbr0bfMnWH:-M2A4y2v6L2-LQyfDx-k:b-4h74ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/11449b96430bde899bc58c028df6cde9f0fc2807/pkg/ui/src/views/app/containers/timewindow/index.tsx#L113)):*
> ```XML
> 
>   componentDidUpdate(prevProps: TimeWindowManagerProps) {
>     if (!_.isEqual(prevProps, this.props)) {
> ```

if this necessary on *any* props change, why add the check?
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-M2A7Wxu4LOFADYCPy9K:-M2A7Wxv7IGAKZYUhcaS:bikbpfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/11449b96430bde899bc58c028df6cde9f0fc2807/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L52)):*
> ```XML
> 
> export const getTimeLabel = (currentWindow?: timewindow.TimeWindow, windowSize?: moment.Duration) => {
>   let time = moment.duration(10, "minutes");
> ```

Why was this added in here? What is the scenario this is accounting for?
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-M2A7kTK4G1CntYDTNiU:-M2A7kTK4G1CntYDTNiV:b-bknurf) ([raw file](https://github.com/cockroachdb/cockroach/blob/11449b96430bde899bc58c028df6cde9f0fc2807/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L223)):*
> ```XML
>   setDatesByQueryParams = (dates?: timewindow.TimeWindow) => {
>     const currentWindow = _.clone(this.props.currentWindow);
>     const end = dates.end || currentWindow && currentWindow.end || moment();
> ```

It's unclear what is happening here. Can you clarify the scenario?
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-M2A7qwW7DMnsB3-vb0d:-M2A7qwW7DMnsB3-vb0e:b-qc6xvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/11449b96430bde899bc58c028df6cde9f0fc2807/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L285)):*
> ```XML
>     const { currentWindow } = this.props;
>     const differenceEndToNow = currentWindow && moment.duration(moment().diff(currentWindow.end)).asMinutes();
>     const differenceEndToStart = currentWindow && moment.duration(moment(currentWindow.end).diff(currentWindow.start)).asMinutes();
> ```

Why does this check need to be added? Does it make sense to just return null if `currentWindow` isn't set?
I'd rather have logic that fails the function early instead of setting bad values on variables.


<!-- Sent from Reviewable.io -->

vladlos committed:
ui: unsafe lifecycle methods

changed `componentWillMount` to `componentDidMount` and `componentWillReceiveProps` to `componentDidUpdate`.
fixed broken code parts due to change lifecycle methods

Resolves: #45189

Release note (ui): none

elkmaster commented:
change `componentWillMount` to `componentDidMount` is pretty painless, nothing relevant changes from development perspective as `componentWillMount` run async, so in real life is mostly equals to `componentDidMount`. And in general very similiar situation with `willRecieve` and `didUpdate`. Thats why all `will` methods is being deprecated, as they not working as intended and can lead to unexpected behaviour.

isEqual checks added to avoid unneeded rerenders. due to `willRecieve` runs before render, it wasn't a case (but still was a performance issue), with didUpdate we got real picture and render loops in some cases.

currentWindow, windowSize both optional on redux level. and we can't set some default value, as it should be now(). that's why there is some additional checks for that added. As on first DID mount they not calculated yet.

nathanstilwell commented on pkg/ui/src/views/cluster/containers/timescale/index.tsx:
```diff
@@ -49,7 +49,12 @@ interface TimeScaleDropdownProps extends RouteComponentProps {
+    time = windowSize;
+  } else if (currentWindow) {
+    time = moment.duration(moment(currentWindow.end).diff(currentWindow.start));
+  }
```
I don't care for this change as it introduces mutability. The code in this case is fine, but it set a bad precedent and time is now able to mutable later in the code in unexpected ways. I assume this change was made for readibiltiy, and having it on multiple lines is indeed more readable, but I would keep the ternary statement and spread that across multiple lines instead.

```suggestion
  const time = windowSize ? 
  windowSize : 
  (currentWindow) ?
  moment.duration(moment(currentWindow.end).diff(currentWindow.start)) :
  moment.duration(10, "minutes");
```

The formatting here is perhaps up for debate (I would prefer to defer to [ESLint](https://eslint.org/docs/rules/multiline-ternary)), but I would like to keep the code immutable unless strictly necessary.

nathanstilwell reviewed (changes requested):
I just have one small change request that is unrelated to the lifecycle methods. The lifecycle changes all look harmless to me and get a 👍

dhartunian reviewed (changes requested):
Thanks for the additional context @elkmaster. Once you address @nathanstilwell's comment we can merge this in.

And please add release justification :angel: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-:-M3S7G2f5VX8qv_BAVNj:b-62xyn4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @elkmaster)



<!-- Sent from Reviewable.io -->

vladlos committed:
ui: ternary statement

changed `if` statements in `getTimeLabel` to ternary statement

Release justification: bug fixes and low-risk updates to new functionality

Release note (ui): none

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45947#-:-M3ffrk8789hBOokZ7S5:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @elkmaster)



<!-- Sent from Reviewable.io -->

nathanstilwell reviewed (approved):
LGTM 👍

dhartunian commented:
bors r+
